### ISSUE TITLE
Item cull mask on Light2D not working as expected

### ISSUE THREAD

**Issue by odynm** [godotengine/godot#25992]
**Godot version:**
3.1 beta 5
(3.0.x tested too and has the same issue)

**OS/device including version:**
Windows 10 64, Intel HD Graphics.
Both on GLES2 and GLES3.

**Issue description:**
Set a light 2D item cullmask to Layer 2 only, but it doesn't cast shadow on a LightOccluder2D which light mask's is set to Layer 2 only. 

If both are Layer 1 only, everything works as expected.
If Layer 1 and Layer 2 are checked in both, eveything works as expected.

**Steps to reproduce:**
- A Light 2D with shadow_item_cull_mask set to Layer 2 only;
- A LightOccluder2D with light_mask set to Layer 2 only;
- Light 2D with a standard light.png set to Add, or any other mode.

**Minimal reproduction project:**
[LighTest.zip](https://github.com/godotengine/godot/files/2873514/LighTest.zip)

**Comment by hedin-hiervard** [comment]
It still doesn't work. How do I set the "shadow mask" in LightOccluder2D? I tried with the master branch and it still sets the usual "light_mask" field.

**Comment by Dr4kzor** [comment]
This should be reopen as it still does not work.  Using a light2d on a layer 2 and setting some obj to be on this exact same layer 2 fails. but if the same obj is on layer 1 and light is on layer 2 only it works again! It seems like the light ignores the layer its on and assumes its always 1.

**Comment by Dr4kzor** [comment]
@akien-mga Is it possible to reopen this issue?

**Comment by akien-mga** [comment]
Indeed, the project in OP seems not to work in 3.1.1-stable nor the current master branch (3aa5b54). CC @JFonS

**Comment by JFonS** [comment]
I played a bit with the OP's project. The whole layer system for lights is a bit complicated, and we can probably revise it for 4.0 or later versions.

In the reproduction project the background sprite is not receiving any shadows because it's only on the first light layer while the `Light2D` node has `shadow_item_cull_mask` set to layer 2 only. Adding the background sprite to layer 2 seems to work and the shadow is visible on the background.

As I said, it's a pretty convoluted system and may need better documentation or some improvements to make it easier to use, but so far seems to be working as intended.

**Comment by akien-mga** [comment]
Thanks, closing again then. If another still reproduces issues after taking the above into account, please upload a project that we can test with.

**Comment by jancnahn** [comment]
@akien-mga This is still not working as expected.

I thought it was related to shadow, but it seems to be more to light mask layers.

**Steps to reproduce**
> Place a Light2D node in a scene with a texture. Range -> Item Cull Mask is 1 by default.
> Place a Sprite node in the scene. By default, the visibility Light Mask is 1, and the sprite is lit by the light 2d node.
> Change the sprite's visibility light mask to anything other than 1. The sprite is still lit by the light 2d node.

This happens even when none of the light masks were checked for the sprite.

**Comment by akien-mga** [comment]
This might actually be a duplicate of #39133 which was fixed by #39162 (CC @lawnjelly).

**Comment by Winterslice** [comment]
This is still an issue in 3.2.3. You can recreate it easily in the 2d lighting demo project. Any Light2d that isn't using layer 1 will not cast shadows.

**Comment by lawnjelly** [comment]
I've worked out why the shadow isn't showing, and I actually now suspect this is intended behaviour rather than a bug.

## Explanation
Addressing the situation in the minimum reproduction project where Light2D lightmask is set to layer 2 only, and the light occluder is set to layer 2 for light mask:

* Although the shadow is not showing, the shader _IS_ being drawn to the shadow buffer, i.e. the bitwise logic is working correctly
* The reason the shadow is not showing is not because of this, it is because there is a second mask logic check on the object that is being _DRAWN TO_.

In this case, the object we want to see the shadow on is the background (`bg`). And if you look at the `bg` light mask, it is set to layer 1 only. Therefore the shadow is not appearing, because there is a check between the light shadow mask (layer 2 only) and the bg light mask (layer 1 only).

If you set the `bg` light mask to layer 1 AND layer 2, the shadow appears!! :grin: 

## TLDR

There are TWO checks in involved:
1) To decide whether to draw from the light to the shadow map, determined by the light's lightmask and the occluder's lightmask.
2) To decide whether to draw from the shadow map onto an object, determined y the light shadow mask and the object the shadow would cast onto.
